radicle-node/src/service.rs

```diff
@@ -891,14 +915,46 @@ where
                         match message.is_fresh(&self.storage) {
                             Ok(is_fresh) => {
                                 if is_fresh {
-                                    // TODO: Only fetch if the refs announced are for peers we're tracking.
-                                    self.fetch(message.rid, announcer);
+                                    match repo_policy.scope {
+                                        tracking::Scope::Trusted => {
+                                            // Check every node and only fetch if we're tracking the node.
+                                            let doc = self.get(message.rid)?;
```

Comment:
Yeah that makes sense. Once we have multi-step fetch, we'd only fetch the delegates I guess.

This makes me wonder though: should ref announcements trigger full clones? Currently they do, but it's a mechanism designed to notify peers of ref changes, not of inventory. I think if we want to preserve the current behavior, then we should indeed fetch here -- probably a good idea to confirm current behavior with a test though.

But thinking about this more, we may want to only do updates when receiving ref announcements, to simplify the code.

Reply:
Not exactly: I'm making the distinction between a full clone, and a fetch that updates refs on an existing repo. Ref announcements are meant to trigger the latter, though it's "ok" to use it to trigger full clones as well. But we might be able to simplify our code if we just ignore ref announcements for repos we don't have yet, and let other parts of the code handle that, eg. inventory announcements or track-repo commands. Just a thought.

With regards to fetching refs that changed, I believe git does a good enough job with that, that we don't need to go crazy granular on the fetchspec, as long as we don't fetch namespaces we aren't tracking.

Reply:
> Not exactly: I'm making the distinction between a full clone, and a fetch that updates refs on an existing repo. Ref announcements are meant to trigger the latter, though it's "ok" to use it to trigger full clones as well. But we might be able to simplify our code if we just ignore ref announcements for repos we don't have yet, and let other parts of the code handle that, eg. inventory announcements or track-repo commands. Just a thought.

Mmmmm I see what you mean. I guess my question for this is: when do inventory announcements happen? Like what I track a repository to clone it and my client sees a refs announcement -- it can clone at that stage since it heard about a node that has what I want. Otherwise, it has to wait for another inventory announcement.

Reply:
Yeah, I guess the way to think about it is: how do I go from knowing about a repo, to having it locally? Assuming the default tracking policy, I would send `Command::TrackRepo`, right? So when that command is received from the service, it should check the routing table and fetch it from some node.

Now it's not always that simple, for two reasons:
1. What happens if there are no seeds for that RID in the routing table?
2. What happens if you restart your node after calling TrackRepo?

For (1), what will cause us to have a seed in the routing table is an inventory-ann, so waiting for that would solve the problem. The other way is via a ref-announcement actually. We do update the routing table on ref-ann. But the point here is that the fetched would be triggered by the routing table update, not the ref-announcement. It's a subtle difference I guess, but it keeps the logic simpler.

For (2) it seems like we need some kind of check on node startup (?) that checks whether we have our tracked repos locally, or whether they should be fetched still. So eg. loop through tracked repos, for each that isn't in storage, try to fetch from seed.

Reply:
> If the `Doc` is `None` that means we've never fetched this repository before. So I think we need to line up a `fetch` with `Namespaces::All`.

I could add that, but I wanted to understand, are we supposed to receive `refs_announcement` before we have the repo doc?  Maybe I missed, but it is not clear to me if that's a valid path from reading the above. i.e. whether we should report an error, or simply handle it and fetch all?

Reply:
We *may* receive it, but as mentioned above, I'm not 100% sure it's the right thing to do. However, since that appears to be the previous behavior, we should probably preserve it, and do a full fetch to not break existing tests. At the very least we should choose one or the other way deliberately.

Reply:
In the latest update, I chose the following behavior: if repo didn't exist and in `scope::Trusted`, we will treat the remote of the `ref_announcement` a trusted peer. Then fetch from them, only for the namespace of this remote.

If that does not make sense, I'm okay to always fetch `Namespaces::All`.

Reply:
I think that works if it isn't breaking existing tests. We'll have a chance to re-evaluate this with fintan's multi-step fetch that is in progress.

Reply:
> I think that works if it isn't breaking existing tests.

Are you saying, for example, you don't want to have changes in tests like this one? : https://github.com/radicle-dev/heartwood/blob/68f7c6e83342687c54c28104be958cad8570dfb1/radicle-cli/tests/commands.rs#L499-L506

Reply:
Meantime, I will assume that we don't want any changes of the existing tests, and I will reverse some changes of `namespaces_for` function, i.e. for `scope::Trusted`, if the repo didn't exist yet, we will fetch using `Namespaces::All`. I will keep the changes minimum so we can move this PR forward.

Reply:
I've updated the diff to keep the existing behavior with regarding `namespaces_for`, and hence reduced the overall changes. Please take a look.